csharp_dotnetcore/14.NLP-With-Dispatch/NlpDispatch/NlpDispatchBot.cs

```diff
@@ -0,0 +1,182 @@
+        public NlpDispatchBot(BotServices services)
+        {
+            _services = services ?? throw new System.ArgumentNullException(nameof(services));
+
+            if (!_services.QnAServices.ContainsKey(QnAMakerKey))
+            {
+                throw new System.ArgumentException($"Invalid configuration.  Please check your '.bot' file for a QnA service named '{DispatchKey}'.");
+            }
+
+            if (!_services.LuisServices.ContainsKey(HomeAutomationLuisKey))
+            {
+                throw new System.ArgumentException($"Invalid configuration.  Please check your '.bot' file for a Luis service named '{HomeAutomationLuisKey}'.");
+            }
+
+            if (!_services.LuisServices.ContainsKey(WeatherLuisKey))
+            {
+                throw new System.ArgumentException($"Invalid configuration.  Please check your '.bot' file for a Luis service named '{WeatherLuisKey}'.");
+            }
+        }
+
+        /// <summary>
+        /// Every Conversation turn for our NLP Dispatch Bot will call this method.
+        /// There are no dialogs used, since it's "single turn" processing, meaning a single
+        /// request and response, with no stateful conversation.
+        /// </summary>
+        /// <param name="context">A <see cref="ITurnContext"/> containing all the data needed
+        /// for processing this conversation turn. </param>
+        /// <param name="cancellationToken">(Optional) A <see cref="CancellationToken"/> that can be used by other objects
+        /// or threads to receive notice of cancellation.</param>
+        /// <returns>A task that represents the work queued to execute.</returns>
+        public async Task OnTurnAsync(ITurnContext context, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            if (context.Activity.Type == ActivityTypes.Message && !context.Responded)
+            {
+                // Get the intent recognition result
+                var recognizerResult = await _services.LuisServices[DispatchKey].RecognizeAsync(context, cancellationToken);
+                var topIntent = recognizerResult?.GetTopScoringIntent();
+
+                if (topIntent == null)
+                {
+                    await context.SendActivityAsync("Unable to get the top intent.");
+                }
+                else
+                {
+                    if (topIntent.Value.score < 0.3)
+                    {
+                        await context.SendActivityAsync("I'm not very sure what you want but will try to send your request.");
+                    }
+
+                    await DispatchToTopIntentAsync(context, topIntent);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Depending on the intent from Dispatch, routes to the right Luis model or Qna service.
+        /// </summary>
+        private async Task DispatchToTopIntentAsync(ITurnContext context, (string intent, double score)? topIntent)
+        {
```

Comment:
A tuple? Wow. I don't see those very often as parameters.

Reply:
I can make this 2 parameters and keep it simple. I wanted to provide a sense of grouping, but I could do that as well bu calling them topIntent and topIntentScore for example. Thoughts?

Reply:
I think grouping by name is easier for people. I suspect the vast majority of C# developers don't know what Tuples are,. 

---
In reply to: [215328459](https://github.com/Microsoft/BotBuilder-Samples/pull/319#discussion_r215328459) [](ancestors = 215328459)